File: proga/1_semestr/labs/lab3/src/main/java/dev/lysmux/lab3/entity/ray/Ray.java
package dev.lysmux.lab3.entity.ray;

import dev.lysmux.lab3.common.Case;
import dev.lysmux.lab3.common.HasCase;
import dev.lysmux.lab3.common.Pluralable;

import java.util.Objects;

public class Ray implements Pluralable, HasCase {


    protected VisibilityType visibility;
    protected RayKind kind;

    public Ray(RayKind kind, VisibilityType visibility) {
        this.visibility = visibility;
        this.kind = kind;
    }

    public Ray(RayKind kind) {
        this(kind, VisibilityType.NONE);
    }

    public VisibilityType getVisibility() {
        return visibility;
    }

    public void setVisibility(VisibilityType visibility) {
        this.visibility = visibility;
    }

    public RayKind getKind() {
        return kind;
    }

    public void setKind(RayKind kind) {
        this.kind = kind;
    }

    @Override
    public String getPluralName() {
        return "лучи";
    }

    @Override
    public String caseDeclension(Case to_case) {
        return switch (to_case) {
            case NOMINATIVE, ACCUSATIVE -> "луч";
            case GENITIVE -> "луча";
            case DATIVE -> "лучу";
            case CREATIVE -> "лучом";
            case PREPOSITIONAL -> "луче";
        };
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Ray ray)) return false;
        if (!super.equals(o)) return false;
        return visibility == ray.visibility && kind == ray.kind;
    }

    @Override
    public int hashCode() {
        return Objects.hash(super.hashCode(), visibility, kind);
    }

    @Override
    public String toString() {
        return "луч";
    }
}

File: proga/1_semestr/labs/lab3/src/main/kotlin/dev/lysmux/lab3/entity/ray/Ray.kt
package dev.lysmux.lab3.entity.ray

import dev.lysmux.lab3.common.Case
import dev.lysmux.lab3.common.HasCase
import dev.lysmux.lab3.common.Pluralable

class Ray(
    val kind: RayKind,
    val visibility: VisibilityType
) : Pluralable, HasCase{

    constructor(kind: RayKind) : this(kind, VisibilityType.NONE)

    override fun getPluralName() = "лучи"

    override fun toString() = "луч"

    override fun caseDeclension(toCase: Case): String {
        return when(toCase) {
            Case.NOMINATIVE -> "луч"
            Case.GENITIVE -> "луча"
            Case.DATIVE -> "лучу"
            Case.ACCUSATIVE -> "луч"
            Case.CREATIVE -> "лучом"
            Case.PREPOSITIONAL -> "луче"
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Ray) return false

        if (kind != other.kind) return false
        if (visibility != other.visibility) return false

        return true
    }

    override fun hashCode(): Int {
        var result = kind.hashCode()
        result = 31 * result + visibility.hashCode()
        return result
    }
}